File: 2021/06-lanternfish/06.hs
import Control.Lens
import Data.List.Lens
import Data.List.Split

main = readFile "input.txt" >>= print . solution

solution = totalFish . (!!256) . progression . initialFish . map read . splitOn ","

initialFish init = foldl (flip addFish) (take 7 $ repeat 0) init

addFish f = element f %~ (+1)

iteration ((x:xs), (w:ws)) = (xs ++ [x + w], ws ++ [x])

progression initial = iterate iteration (initial, [0, 0])

totalFish (fish, waiting) = sum (fish ++ waiting)

File: 2021/06-lanternfish/06.py
file = open("input.txt")
contents = file.read().strip()
file.close()

initial = map(int, contents.split(","))

#  fish = list(map(int, contents.split(",")))

#  for day in range(256):
    #  for idx in range(len(fish)):
        #  if fish[idx] == 0:
            #  fish.append(8)
            #  fish[idx] = 6
        #  else:
            #  fish[idx] -= 1


fish = [0] * 7
waiting = [0,0]

for f in initial:
    fish[f] += 1

for day in range(256):
    to_hatch_today = fish[day % 7]
    fish[day % 7] += waiting[0]
    waiting[0] = waiting[1]
    waiting[1] = to_hatch_today

print(sum(fish) + sum(waiting))

